internal/dataapi/server/server.go

```diff
@@ -49,110 +50,157 @@ func New(l *slog.Logger, handler *handler.Handler) *Server {
 type Server struct {
 	l       *slog.Logger
 	handler *handler.Handler
+
+	m sync.Map
 }
 
-// AuthMiddleware handles SCRAM authentication based on the username and password specified in request.
-// After a successful handshake it calls the next handler.
+// AuthMiddleware authenticates the request using bearer token or basic authentication.
+// If bearer token is provided, it authenticates using bearer token.
+// Otherwise, basic auth is used and upon successful authentication
+// it sets a bearer token in the response header.
+// After a successful authentication, it calls the next handler.
 func (s *Server) AuthMiddleware(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ctx := r.Context()
-		username, password, ok := r.BasicAuth()
+		if bearer := strings.HasPrefix(r.Header.Get("Authorization"), "Bearer "); bearer {
+			if ok := s.bearerAuth(w, r); !ok {
+				return
+			}
 
-		if !ok {
-			writeError(w, errorNoAuthenticationSpecified, http.StatusBadRequest)
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		if password == "" || username == "" {
-			writeError(w, errorMissingAuthenticationParameter, http.StatusBadRequest)
+		if ok := s.basicAuth(w, r); !ok {
 			return
 		}
 
-		client, err := scram.SHA256.NewClient(username, password, "")
-		if err != nil {
-			http.Error(w, lazyerrors.Error(err).Error(), http.StatusBadRequest)
+		if err := s.setBearerTokenHeader(w); err != nil {
```

Comment:
Alternatively we can follow more closely to OAuth2 for having specific endpoint for token. https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.4